### ISSUE TITLE
AppRegistryNotReady exception

### ISSUE THREAD

**Issue by tsebire** [etianen/django-watson#144]
Hello,

I can't register a model in watson anymore since I've upgraded to Django 1.9.

It raises a AppRegistryNotReady exception. I'm using an AppConfig for my module set in my __init__.py file :

```python
from django.apps import AppConfig
from watson import search as watson

class ProduitsConfig(AppConfig):
    name = "moderation.apps.produits"

    def ready(self):
        ProduitModel = self.get_model("Produit")
        watson.register(ProduitModel, fields=("nom", "marque__nom", "marque__pays__nom", "marque__region__nom", "type", "couleur"))
```

When I comment the watson import and the register line, I don't have any error.

Am I missing something ?

**Comment by etianen** [comment]
Try upgrading to the latest django-watson.


https://github.com/etianen/django-watson/blob/master/CHANGELOG.markdown

**Comment by tsebire** [comment]
I already did, that's why I don't know what can I do next. I'll retry tonight and keep you updated.

**Comment by etianen** [comment]
Hmm, do you have a full traceback that I can investigate?

**Comment by tsebire** [comment]
Here you go :

Traceback (most recent call last):
  File "D:/Work/Sites/moderation/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "C:\Python27\lib\site-packages\django\core\management\__init__.py", line 350, in execute_from_command_line
    utility.execute()
  File "C:\Python27\lib\site-packages\django\core\management\__init__.py", line 342, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "C:\Python27\lib\site-packages\django\core\management\__init__.py", line 176, in fetch_command
    commands = get_commands()
  File "C:\Python27\lib\site-packages\django\utils\lru_cache.py", line 100, in wrapper
    result = user_function(*args, **kwds)
  File "C:\Python27\lib\site-packages\django\core\management\__init__.py", line 71, in get_commands
    for app_config in reversed(list(apps.get_app_configs())):
  File "C:\Python27\lib\site-packages\django\apps\registry.py", line 137, in get_app_configs
    self.check_apps_ready()
  File "C:\Python27\lib\site-packages\django\apps\registry.py", line 124, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.

**Comment by tsebire** [comment]
I fixed the problem by putting the watson import in the ready method of AppConfig and not at the beginning of my file (it was working like this before).

Thanks for your time.

**Comment by fresh-django** [comment]
@tsebire thanks for posting your fix. I'd been looking at this for about 2 hours now!

**Comment by etianen** [comment]
Out of interest, what causes the exception? It is running buildwatson, or
is it runserver?

I'm confused, because none of the unit tests pick up on this.

**Comment by tsebire** [comment]
It was thrown on runserver command.

**Comment by etianen** [comment]
I'd like to keep this open for now. It should be possible to import watson at the top-level of your apps.py file. I'm not sure right now why it's not...

**Comment by tsebire** [comment]
Ok my bad :). Good luck fixing this and happy new year.

**Comment by orf** [comment]
I'm also seeing this with Django 1.9.1:

```
from django.apps import AppConfig
from watson import search as watson

default_app_config = 'issues.IssueAppConfig'


class IssueAppConfig(AppConfig):
    name = "issues"

    def ready(self):
        software_v = self.get_model("SoftwareVersion")
        issue = self.get_model("IssueTranslation")
        watson.register(software_v)
        watson.register(issue)
```

Throws: 
```django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.```

**Comment by Athemis** [comment]
What worked for me is to move "from watson import search as watson" into the ready() function. No idea what's going on behind the scene though.

**Comment by etianen** [comment]
Check out the latest master branch. This should be fixed.

**Comment by etianen** [comment]
Keeping this issue open for now, until confirmation that it's fixed...

**Comment by acer9997** [comment]
@Athemis moving into ready() worked for me.

**Comment by TommySprat** [comment]
@etianen I tried it with the latest master branch. Unfortunately it still gives the same error:
`django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.`
Moving the imports into the ready function still fixes it.

**Comment by etianen** [comment]
Can I have the full traceback?

**Comment by TommySprat** [comment]
```
Unhandled exception in thread started by <function check_errors.<locals>.wrapper at 0x035E1B70>
Traceback (most recent call last):
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\core\management\commands\runserver.py", line 109, in inner_run
    autoreload.raise_last_exception()
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\autoreload.py", line 249, in raise_last_exception
    six.reraise(*_exception)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\apps\registry.py", line 85, in populate
    app_config = AppConfig.create(entry)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\apps\config.py", line 90, in create
    module = import_module(entry)
  File "C:\Program Files (x86)\Python35-32\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 944, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 662, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "E:\Playground\Python\CaloriesNow\foodlog\apps.py", line 2, in <module>
    from watson import search as watson
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\watson\search.py", line 14, in <module>
    from django.contrib.contenttypes.models import ContentType
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\contrib\contenttypes\models.py", line 159, in <module>
    class ContentType(models.Model):
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\models\base.py", line 94, in __new__
    app_config = apps.get_containing_app_config(module)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\apps\registry.py", line 239, in get_containing_app_config
    self.check_apps_ready()
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\apps\registry.py", line 124, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
```

**Comment by etianen** [comment]
Oh, the ContentTypes framework. Gotcha.

Try the latest master now.

**Comment by sunjoomoon** [comment]
I had the same issue but the issue RESOLVED as per Dave's advice to use the latest master branch. Thanks for the hard work.

**Comment by etianen** [comment]
Cool, I'll release on Monday! :D

**Comment by TommySprat** [comment]
I also tried that latest master branch and can confirm that it's fixed on my system too.

**Comment by etianen** [comment]
This is now released as 1.2.1
